### PR TITLE
Add support for recursive delete in EncryptedFileSystem

### PR DESCRIPTION
**Pull request by TheBeruriahIncident**
Fixes https://github.com/palantir/hadoop-crypto/issues/117

We want to be able to delete folders of many items. CC @eraboin

If this change seems generally reasonable, I can add more tests

### CODE REVIEW

File: hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/EncryptedFileSystem.java [hunk 1]
```diff
@@ -143,8 +144,16 @@ private void tryRemoveKey(Path path) {
...
-        if (recursive) {
-            throw new UnsupportedOperationException("EncryptedFileSystem does not support recursive deletes");
+        if (recursive && isDirectory(path)) {
+            FileStatus[] children = listStatus(path);
+
+            for (FileStatus child : children) {
+                boolean deleteSuccessful = delete(child.getPath(), true);
+
+                if (!deleteSuccessful) {
+                    return false;
```
Comment (eraboin) [inline hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/EncryptedFileSystem.java:154]: The definition of `delete` in the [Hadoop documentation](http://hadoop.apache.org/docs/r2.7.5/api/org/apache/hadoop/fs/FilterFileSystem.html#delete-org.apache.hadoop.fs.Path-boolean-) does not specify whether we should stop immediately here, or continue with the rest of the delete. If we're modeling this after `rm -rf` then I'd expect it to continue deleting what it can though, right?

Not sure if there's a reference impl elsewhere we can look at.

Comment (TheBeruriahIncident) [reply hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/EncryptedFileSystem.java:154]: Easy to do either, but yeah, I'm unsure what behavior is preferable. RawLocalFileSystem continues and eventually returns false if anything failed. FTPFileSystem continues and ignores any intermediate issues (maybe a bug). DistributedFileSystem is hard to follow, but seems to atomically destroy everything, making this is a non-issue. S3FileSystem matches the behavior of this PR, returning early if anything fails.

Comment (ellisjoe) [reply hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/EncryptedFileSystem.java:154]: I'd really prefer to avoid supporting recursive deletes within EncryptedFileSystem until we can do so efficiently. I'd propose that you do:
```
RemoteIterator<LocatedFileStatus> files = efs.listFiles(path, true);
while (files.hasNext()) {
    efs.delete(files.next().getPath(), false);
}
```
as a workaround for the time being.
